File: concepts/graph/graph_traverse/DFS.py
from enum import Enum


class Colors(Enum):
    WHITE = 0
    GRAY = 1
    BLACK = 2


def DFS(vertex, graph, color, parent, result):

    color[vertex] = Colors.GRAY
    result.append(vertex)

    for neighbor in graph[vertex]:

        if color[neighbor] == Colors.WHITE:  

            parent[neighbor] = vertex
            DFS(neighbor, graph, color, parent, result)

    color[vertex] = Colors.BLACK


if __name__ == "__main__":
    n = 6
    graph = [[] for _ in range(n)]
    graph[0] = [1, 2]
    graph[1] = [0, 3, 4]
    graph[2] = [0, 4]
    graph[3] = [1, 5]
    graph[4] = [1, 2, 5]
    graph[5] = [3, 4]

    color = [Colors.WHITE] * n
    parent = [-1] * n
    result = []

    DFS(0, graph, color, parent, result)

    print("DFS Order:", result)
    print("Parents:", parent)

File: concepts/graph/graph_traverse/DFS.cpp
#include <bits/stdc++.h>
using namespace std;

enum Color
{
    WHITE,
    GRAY,
    BLACK
};

void DFS(int u, vector<vector<int>> &adj, vector<Color> &color, vector<int> &parent, vector<int> &result)
{
    color[u] = GRAY;
    result.push_back(u);

    for (int v : adj[u])
    {
        if (color[v] == WHITE)
        {
            parent[v] = u;
            DFS(v, adj, color, parent, result);
        }
    }

    color[u] = BLACK;
}

int main()
{
    int n = 6;
    vector<vector<int>> adj(n);

    adj[0] = {1, 2};
    adj[1] = {0, 3, 4};
    adj[2] = {0, 4};
    adj[3] = {1, 5};
    adj[4] = {1, 2, 5};
    adj[5] = {3, 4};

    vector<Color> color(n, WHITE);
    vector<int> parent(n, -1);
    vector<int> result;

    DFS(0, adj, color, parent, result);

    cout << "DFS Order:\n";
    for (int v : result)
        cout << v << " ";
    cout << "\nParents:\n";
    for (int i = 0; i < n; i++)
        cout << i << " -> " << parent[i] << "\n";
}
